azavea/raster-vision
# Add support for nvidia-docker2 with Deep Learning Base AMI

## Overview

Add support for [nvidia-docker2](https://github.com/NVIDIA/nvidia-docker/wiki/About-version-2.0) with [Deep Learning Base AMI](https://docs.aws.amazon.com/batch/latest/userguide/batch-gpu-ami.html)

* Update driver.py AMI search regex
* Add nvidia-docker2 configuration scripts

Connects raster-foundry/prediction-api#59 and #152 
Resolves #233 

## Notes

The AMI that `driver.py` was returning is outdated, so I fixed it. You may need to update the regex again if you want to use the new Deep Learning AMI with preinstalled ML libraries. The Deep Learning Base AMI is just Amazon Linux optimized for GPU workloads with NVIDIA/CUDA drivers preinstalled.

Also, we are publishing identical AMIs in Raster Foundry that I have successfully tested with AWS Batch on `p3.2xlarge` instances. I did not investigate anything with Raster Vision deployment in #152, because that wasn't in the scope of the changes we made with raster-foundry/prediction-api#59.

nvidia-docker2 configures itself as a Docker runtime. We don't need to add volume mounts, etc anymore, everything should just work. If you have any questions, I have experience with this workflow on AWS Batch with Raster Foundry.

## Testing Instructions

First, I brought up a dev environment:

```bash
$ mkdir ~/raster_vision_data && export RASTER_VISION_DATA_DIR=~/raster_vision_data/
$ ./scripts/setup
$ vagrant ssh
```

Then, run `batch_amis.py`, and make sure it finds the latest Deep Learning Base AMI:

```bash
vagrant@raster-vision:/vagrant$ cd deployment/ && ./batch_amis.py  build-amis --aws-profile raster-vision
Creating GPU Workload AMI in us-east-1 region
Found AMI: ami-a8c5c0d7
Running Packer Command: packer build -var raster_vision_gpu_version=d643ca2 -var aws_region=us-east-1 -var aws_gpu_ami=ami-a8c5c0d7 -var branch=feature/jrb/add-p3-packer /vagrant/deployment/packer/template.js
...
```

Ensure that the `nvidia-docker2` installation was successful:

```bash
...
    raster-vision-gpu: +-----------------------------------------------------------------------------+
    raster-vision-gpu: | NVIDIA-SMI 396.37                 Driver Version: 396.37                    |
    raster-vision-gpu: |-------------------------------+----------------------+----------------------+
    raster-vision-gpu: | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
    raster-vision-gpu: | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
    raster-vision-gpu: |===============================+======================+======================|
    raster-vision-gpu: |   0  Tesla V100-SXM2...  On   | 00000000:00:1E.0 Off |                    0 |
    raster-vision-gpu: | N/A   39C    P0    26W / 300W |      0MiB / 16160MiB |      0%      Default |
    raster-vision-gpu: +-------------------------------+----------------------+----------------------+
...
```

Finally, ensure the AMI is built:

```bash
...
==> Builds finished. The artifacts of successful builds are:
--> raster-vision-gpu: AMIs were created:

us-east-1: ami-069ba3044e1b1f4ea
```
